File: MATLAB/Solvers/GASSO/GASSO.m
%==========================================================================
%                        The GASSO Algorithm 
%==========================================================================
% DATE
%        February 2018
%        edited December 2019
%
% AUTHOR
%        Sait Cakmak
%        scakmak3 AT gatech DOT edu
%        Enlu Zhou
%        enlu.zhou AT isye DOT gatech DOT edu
%
% REFERENCE
%        Enlu Zhou, Shalabh Bhatnagar (2018) 
%        Gradient-Based Adaptive Stochastic Search 
%               for Simulation Optimization Over Continuous Space. 
%        INFORMS Journal on Computing 30(1):154-167. 
%        https://doi.org/10.1287/ijoc.2017.0771
%==========================================================================
% 
% INPUT 
%        probHandle
%              Problem function name
%        probstructHandle
%              Problem structure function name 
%        problemRng
%              Random number generators (streams) for problems
%        solverRng
%              Random number generator (stream) for solver
%
% OUTPUT
%        Ancalls
%              An array (size = 'NumSoln' X 1) of budget expended 
%        A 
%              An array (size = 'NumSoln' X 'dim') of solutions
%              returned by solver
%        AFn   
%              An array (size = 'NumSoln' X 1) of estimates of expected 
%              objective function value
%        AFnVar
%              An array of variances corresponding to  
%              the objective function at A
%        AFnGrad
%              An array of gradient estimates at A; not reported
%        AFnGradCov
%              An array of gradient covariance matrices at A; not reported
%        AConstraint
%              A vector of constraint function estimators; not applicable
%        AConstraintCov
%              An array of covariance matrices corresponding to the
%              constraint function at A; not applicable
%        AConstraintGrad
%              An array of constraint gradient estimators at A; not
%              applicable
%        AConstraintGradCov
%              An array of covariance matrices of constraint gradient  
%              estimators at A; not applicable
%
%==========================================================================

%% GASSO
function [Ancalls, A, AFnMean, AFnVar, AFnGrad, AFnGradCov, ...
     AConstraint, AConstraintCov, AConstraintGrad, ...
     AConstraintGradCov] = GASSO(probHandle, probstructHandle, ...
    problemRng, solverRng)

% not reported
AFnGrad = NaN;
AFnGradCov = NaN;
AConstraint = NaN;
AConstraintCov = NaN;
AConstraintGrad = NaN;
AConstraintGradCov = NaN;

% Separate the two solver random number streams
solverInitialRng = solverRng{1}; % RNG for finding initial solutions
solverInternalRng = solverRng{2}; % RNG for the solver's internal randomness

%% Gather parameters
NumStartingSols = 100;

% Get problem dimension and bounds and initial solution(s)
RandStream.setGlobalStream(solverInitialRng)
[minmax, dim, ~, ~, VarBds, ~, ~, ssolsM, budget, ~, ~, ~] = probstructHandle(NumStartingSols);

% Set parameters
N = round(50 * sqrt(dim)); % number of samples for each iteration
M = 15; % each candidate solution is simulated M times
K = floor(budget./(N*M)); % max number of iterations
rho = 0.15; % quantile parameter

% step size - alpha_k = alpha_0 / (k + alpha_c)^alpha_p
alpha_0 = 15; % step size numerator default = 15
alpha_c = 150; % step size denominator constant default = 150
alpha_p = 0.6; % step size denominator exponent default = 0.6

% Determine maximum number of solutions that can be reported within max budget
MaxNumSoln = K + 2; % + 1 for initial solution, + 1 for final solution

% Initialize
Ancalls = zeros(MaxNumSoln, 1);
A = zeros(MaxNumSoln, dim);
AFnMean = zeros(MaxNumSoln, 1);
AFnVar = zeros(MaxNumSoln, 1);

% Using CRN: for each solution, start at substream 1
problemseed = 1;

% Record initial solution (first one in ssolsM)
Ancalls(1) = 0;
A(1,:) = ssolsM(1,:);
AFnMean(1) = NaN; % unevaluated
AFnVar(1) = NaN; % unevaluated

% Average initial solutions
x0 = mean(ssolsM);

%% The actual algorithm
% GASSO: sampling distribution is normal (for unbounded solution set) or truncated normal (for bounded solution set)
% 
% N: # of samples
% M: # of evaluations per sample
% K: # of iterations
% dim: dimension of the problem
% alpha: step size
% rho: quantile parameter
% samples: number of samples
% Hbar: optimal value estimate
% xbar: candidate solution
% hvar: variance of optimal value estimate
% x0: starting solution if one is asked for, o.w. NaN
% mu: mean of the sampling distribution
% var: variance of the sampling distribution


%% Initialization

% initialize mu_k based on starting solutions
mu_k = x0;

% set the initial search variance based on starting solutions
var_k = var(ssolsM);

% reparameterization for exponential family
theta1_k = mu_k./var_k;
theta2_k = -0.5*ones(1,dim)./var_k;
theta_k = [theta1_k, theta2_k];

% Set random number generator for generating random candidate solutions
RandStream.setGlobalStream(solverInternalRng)

% generate candidate solutions from (truncated) normal distribution
x = zeros(N,dim);
kk = 1;
while kk <= N
    X_k = randn(N,dim)*diag(sqrt(var_k))+ones(N,1)*mu_k;
    for i = 1:N
        % accept / reject samples based on variable bounds
        % can be extended to handle deterministic constraints
        if all(X_k(i,:)' >= VarBds(:,1)) && all(X_k(i,:)' <= VarBds(:,2)) && kk <= N
            x(kk,:) = X_k(i,:);
            kk = kk+1;
        end % if
    end % for
end % while
X_k = x;

k = 1;
samples = zeros(K,1);
Hbar = AFnMean(1); % best value so far. Start with initial solution value
xbar = zeros(K,dim);
hvar = zeros(K,1);
Ntotal = 0;

%% iteration
while k <= K
    alpha_k = alpha_0/(k+alpha_c)^alpha_p; % set step size
   
    H = zeros(N,1); % obj function values
    H_var = zeros(N,1); % obj function variances
    for i = 1:N
        % simulate each solution X_k(i) for M times - uses CRN
        [fn, H_var(i), ~, ~, ~, ~, ~, ~] = probHandle(X_k(i,:), M, problemRng, problemseed);
        result = minmax*fn;
        H(i) = result;
    end %for
    problemseed = problemseed + 1;

    % return an error message and stop the algorithm
    % if the problem returns -inf for feasible solutions
    if all(H == -inf)
        fprintf('The problem returns -inf for valid x values!!!');
        return
    end %if

    Ntotal = Ntotal + N*M;
    samples(k) = Ntotal; % record the budget
    
    % keep the best candidate
    [Hbar(k), I] = max(H);
    hvar(k) = H_var(I);
    xbar(k,:) = X_k(I,:);
    
    if k > 1
        if Hbar(k) < Hbar(k-1) || isnan(Hbar(k)) == 1      
            Hbar(k) = Hbar(k-1);
            xbar(k,:) = xbar(k-1,:);
            hvar(k) = hvar(k-1);
        end %if
    end %if
    
    % Shape function - take elite samples
    [G_sort, ~] = sort(H,'descend');   % sorted performances
    gm = G_sort(ceil(N*rho));
    S_theta = (H > gm);
    
    % estimate gradient and hessian
    w_k = S_theta/sum(S_theta);
    CX_k = [X_k,X_k.*X_k]; 
    grad_k = w_k'*CX_k - [mu_k, var_k + mu_k.*mu_k];  
    Hes_k = -cov(CX_k);        
    Hes_inv_k = (Hes_k + 1e-8.*eye(2*dim))\diag(ones(1,2*dim)); 
    
    % update the parameter using an SA iteration
    theta_k = theta_k - alpha_k*(Hes_inv_k*grad_k')';
    theta1_k = theta_k(1:dim);
    theta2_k = theta_k(dim+1:2*dim);
    var_k = -0.5./theta2_k;
    mu_k = theta1_k.*var_k;
    
    % project mu_k and var_k to feasible parameter space
    for i = 1:dim
        if mu_k(i) < VarBds(i,1)
            mu_k(i) = VarBds(i,1);
        elseif mu_k(i) > VarBds(i,2)
            mu_k(i) = VarBds(i,2);
        end %if
    end %for
    var_k = abs(var_k);

    % Set random number generator for generating random candidate solutions
    RandStream.setGlobalStream(solverInternalRng)
    
    % generate candidate solutions from (truncated) normal distribution
    x = zeros(N,dim);
    kk = 1;
    while kk <= N
        X_k = randn(N,dim)*diag(sqrt(abs(var_k))) + ones(N,1)*mu_k;
        for i = 1:N
            % accept / reject samples based on variable bounds
            % can be extended to handle deterministic constraints
            if all(X_k(i,:)' >= VarBds(:,1)) && all(X_k(i,:)' <= VarBds(:,2)) && kk<=N
                x(kk,:) = X_k(i,:);
                kk = kk + 1;
            end % if
        end % for
    end % while
    X_k = x;

    k = k + 1;
    clear H_var H S_theta w_k theta1_k theta2_k CX_k grad_k Hes_k Hes_inv_k;
end %while

% return solutions
Ancalls(2:K+1) = samples;
A(2:K+1,:) = xbar;
AFnMean(2:K+1) = minmax*Hbar;
AFnVar(2:K+1) = hvar;

% Record solution at max budget
Ancalls(K+2) = budget;
A(K+2,:) = xbar(K,:);
AFnMean(K+2) = minmax*Hbar(K);
AFnVar(K+2) = hvar(K);

end %GASSO

File: simopt/solvers/gasso.py
"""
Summary
-------
Iteratively generates population of candidate solutions from a sample distribution and use its performance
to update the sample distribution.
A detailed description of the solver can be found `here <https://simopt.readthedocs.io/en/latest/gasso.html>`_.
"""
from ..base import Solver
import numpy as np
import warnings
warnings.filterwarnings("ignore")

class GASSO(Solver):
    """
    A solver that iteratively generates population of candidate solutions from a sample distribution 
    and use its performance to update the sample distribution.

    Attributes
    ----------
    name : string
        name of solver
    objective_type : string
        description of objective types:
            "single" or "multi"
    constraint_type : string
        description of constraints types:
            "unconstrained", "box", "deterministic", "stochastic"
    variable_type : string
        description of variable types:
            "discrete", "continuous", "mixed"
    gradient_needed : bool
        indicates if gradient of objective function is needed
    factors : dict
        changeable factors (i.e., parameters) of the solver
    specifications : dict
        details of each factor (for GUI, data validation, and defaults)
    rng_list : list of mrg32k3a.mrg32k3a.MRG32k3a objects
        list of RNGs used for the solver's internal purposes

    Arguments
    ---------
    name : str
        user-specified name for solver
    fixed_factors : dict
        fixed_factors of the solver

    See also
    --------
    base.Solver
    """
    def __init__(self, name="GASSO", fixed_factors=None):
        if fixed_factors is None:
            fixed_factors = {}
        self.name = name
        self.objective_type = "single"
        self.constraint_type = "box" 
        self.variable_type = "continuous"
        self.gradient_needed = True
        self.specifications = {
            "crn_across_solns": {
                "description": "use CRN across solutions?",
                "datatype": bool,
                "default": True
            },
            "sample_size": {
                "description": "sample size per solution",
                "datatype": int,
                "default": 10
            },
            "max_iter": {
                "description": "maximum number of iterations",
                "datatype": int,
                "default": 10000
            },
            "rho": {
                "description": "quantile parameter",
                "datatype": float,
                "default": 0.15
            },
            "M": {
                "description": "times of simulations for each candidate solution",
                "datatype": int,
                "default": 15
            },
            "alpha_0": {
                "description": "step size numerator",
                "datatype": int,
                "default": 15
            },
            "alpha_c": {
                "description": "step size denominator constant",
                "datatype": int,
                "default": 150
            },
            "alpha_p": {
                "description": "step size denominator exponent",
                "datatype": float,
                "default": 0.6
            },
            "MaxNumSoln": {
                "description": "maximum number of solutions that can be reported within max budget",
                "datatype": int,
                "default": 10002
            }
        }
        self.check_factor_list = {
            "crn_across_solns": self.check_crn_across_solns,
            "sample_size": self.check_sample_size,
            "max_iter": self.check_max_iter,
            "rho": self.check_rho,
            "M": self.check_M,
            "alpha_0": self.check_alpha_0,
            "alpha_c": self.check_alpha_c,
            "alpha_p": self.check_alpha_p,
            "MaxNumSoln": self.check_MaxNumSoln
        }
        super().__init__(fixed_factors)

    def check_sample_size(self):
        return self.factors["sample_size"] > 0

    def check_max_iter(self):
        return self.factors["max_iter"] > 0

    def check_rho(self):
        return 0 < self.factors["rho"] < 1

    def check_M(self):
        return self.factors["M"] > 0

    def check_alpha_0(self):
        return self.factors["alpha_0"] > 0

    def check_alpha_c(self):
        return self.factors["alpha_c"] > 0

    def check_alpha_p(self):
        return 0 < self.factors["alpha_p"] < 1

    def check_MaxNumSoln(self):
        return self.factors["MaxNumSoln"] > 0

    def solve(self, problem):
        """
        Run a single macroreplication of a solver on a problem.

        Arguments
        ---------
        problem : Problem object
            simulation-optimization problem to solve

        Returns
        -------
        recommended_solns : list of Solution objects
            list of solutions recommended throughout the budget
        intermediate_budgets : list of ints
            list of intermediate budgets when recommended solutions changes
        """
        dim = problem.dim
        rand_sol_rng = self.rng_list[0]
        x_ini = [problem.get_random_solution(rand_sol_rng) for i in range(9)]
        x_ini.append(problem.factors["initial_solution"])

        # Initialize sampling distribution based on the initial solution population
        mu_k = np.mean(x_ini, axis = 0)  # The mean for each dim
        var_k = np.var(x_ini, axis = 0)
        theta1_k = mu_k/var_k
        theta2_k = -0.5 * np.ones(problem.dim) / var_k
        theta_k = np.append(theta1_k, theta2_k)
        N = int(50 * np.sqrt(dim))
        K = int(np.floor(problem.factors['budget']/(N * self.factors['M'])))
        MaxNumSoln = K + 2

        recommended_solns = []
        intermediate_budgets = []
        expended_budget = 0
        # Designate random number generator for random sampling.
        find_next_soln_rng = self.rng_list[1]

        # Get random solutions from normal distribution (truncated)
        x = np.zeros((N, dim))
        kk = 0
        while kk < N:
            normal_vec = np.array([find_next_soln_rng.normalvariate() for i in range(N * dim)]).reshape((N,dim))
            X_k = np.multiply(normal_vec, np.sqrt(var_k)) + np.ones((N, dim)) * mu_k
            for i in range(N):
                if all(X_k[i, :] >= problem.lower_bounds) and all(X_k[i, :] <= problem.upper_bounds) and kk < N:
                    x[kk, :] = X_k[i, :]
                    kk += 1
        X_k = x

        # Create the initial solution based on the truncated normal
        new_solution = self.create_new_solution(problem.factors['initial_solution'], problem)
        recommended_solns.append(new_solution)
        intermediate_budgets.append(expended_budget)
        problem.simulate(new_solution, self.factors['M'])
        expended_budget += self.factors['M']

        # Track the internal updates
        Hbar = np.zeros(K)
        xbar = np.zeros((K, dim))
        hvar = np.zeros(K)
        k = 0

        # Sequentially generate random solutions and simulate them.
        while expended_budget < problem.factors['budget'] and k < K:
            # Update alpha
            alpha_k = self.factors['alpha_0'] / (k + self.factors['alpha_c']) ** self.factors['alpha_p']
            H = np.zeros(N)
            H_var = np.zeros(N)
            # Sample N new solution candidates by updated distribution
            for i in range(N):
                new_solution = self.create_new_solution(X_k[i, :], problem)
                X_k[i, :] = new_solution.x
                problem.simulate(new_solution, self.factors['M'])
                expended_budget += self.factors['M']
                H[i] = problem.minmax * new_solution.objectives_mean
                H_var[i] = np.var(new_solution.objectives)
            # Find the best one among the N candidates
            Hbar[k], idx = np.max(H), np.argmax(H)
            hvar[k] = H_var[idx]
            xbar[k, :] = X_k[idx, :]
            new_solution = X_k[idx, :]

            if k >= 1:
                # Compare the new best candidate to previous best solution, if better then update
                if Hbar[k] < Hbar[k-1] and Hbar[k] != None:
                    Hbar[k] = Hbar[k-1]
                    xbar[k, :] = xbar[k-1, :]
                    hvar[k] = hvar[k-1]
                    new_solution = xbar[k, :]

            # Track the new candidate and update recommended_sols and budgets
            new_solution = self.create_new_solution(tuple(xbar[k, :]), problem)
            recommended_solns.append(new_solution)
            intermediate_budgets.append(expended_budget)
            # Shape function
            G_sort = np.sort(H)[::-1]
            gm = G_sort[int(np.ceil(N * self.factors['rho']))]
            S_theta = H > gm

            # Estimate gradient and hessian
            w_k = S_theta/sum(S_theta)
            CX_k = np.hstack((X_k, X_k * X_k)).T
            grad_k = np.matmul(w_k.T, CX_k.T) - np.vstack((mu_k, var_k + mu_k * mu_k)).reshape(1, -1)
            Hes_k = -np.cov(CX_k)
            Hes_k_inv = np.linalg.inv(Hes_k + 1e-5 * np.eye(2*dim)) @ np.diag(np.ones(2*dim))
            # Update the parameter using an SA iteration
            theta_k -= (alpha_k * (Hes_k_inv @ grad_k.T)).reshape(1, -1)[0]
            theta1_k = theta_k[:dim]
            theta2_k = theta_k[dim: 2 * dim]
            var_k = -0.5/theta2_k
            mu_k = theta1_k * var_k

            # Project mu_k and var_k to feasible parameter space
            for i in range(dim):
                if mu_k[i] < problem.lower_bounds[i]:
                    mu_k[i] = problem.lower_bounds[i]
                if mu_k[i] > problem.upper_bounds[i]:
                    mu_k[i] = problem.upper_bounds[i]
            var_k = abs(var_k)

            # Generate new candidate solutions from the truncated normal distribution
            x = np.zeros((N, dim))
            kk = 0
            while kk < N:
                normal_vec = np.array([find_next_soln_rng.normalvariate() for i in range(N * dim)]).reshape((N,dim))
                X_k = np.multiply(normal_vec, np.sqrt(var_k)) + np.ones((N, dim)) * mu_k
                for i in range(N):
                    if all(X_k[i, :] >= problem.lower_bounds) and all(X_k[i, :] <= problem.upper_bounds) and kk < N:
                        x[kk, :] = X_k[i, :]
                        kk += 1
            k += 1
            X_k = x

        return recommended_solns, intermediate_budgets